numpy/core/src/multiarray/common.c

```diff
@@ -164,7 +164,7 @@ PyArray_DTypeFromObjectHelper(PyObject *obj, int maxdims,
 
             if (string_type == NPY_STRING) {
                 if ((temp = PyObject_Str(obj)) == NULL) {
-                    return -1;
+                    goto fail;
```

Comment:
This requires *out_dtype to be set to NULL before, which is not guarenteed inside this function (though may be done by the caller). It would be safer if you set it to NULL at the top of the function.

Reply:
Good point, but I think the function requires `out_dtype` to be either `NULL` or a "minimal" dtype. See the comment of the previous function.

Reply:
`out_dtype` can be promoted/changed for strings, so the caller has to be careful I think.